File: ChipTallyiOS/ChipTally/Presentation/Setup/SetupViewModel.swift
//
//  SetupViewModel.swift
//  ChipTally
//

import Foundation

protocol SetupViewModelDelegate: AnyObject {
    func didUpdatePlayerCount(_ count: Int)
    func didValidateForm(_ isValid: Bool)
    func didLoadSavedSettings()
}

final class SetupViewModel {
    weak var delegate: SetupViewModelDelegate?

    private let createGameSessionUseCase: CreateGameSessionUseCaseProtocol
    private let gameRepository: GameRepositoryProtocol
    private let userDefaults: UserDefaults

    private static let playerNamesKey = "lastPlayerNames"
    private static let chipCountKey = "lastChipCount"

    private(set) var playerCount: Int = 2 {
        didSet {
            delegate?.didUpdatePlayerCount(playerCount)
            validateForm()
        }
    }

    private(set) var playerNames: [String] = ["", ""]
    private(set) var initialChipCount: Int = 100

    var minPlayerCount: Int { 2 }
    var maxPlayerCount: Int { 10 }

    init(
        createGameSessionUseCase: CreateGameSessionUseCaseProtocol = CreateGameSessionUseCase(),
        gameRepository: GameRepositoryProtocol = GameRepository(),
        userDefaults: UserDefaults = .standard
    ) {
        self.createGameSessionUseCase = createGameSessionUseCase
        self.gameRepository = gameRepository
        self.userDefaults = userDefaults
    }

    func loadSavedSettings() {
        if let savedNames = userDefaults.stringArray(forKey: Self.playerNamesKey),
           !savedNames.isEmpty {
            playerNames = savedNames
            playerCount = savedNames.count
        }

        let savedChipCount = userDefaults.integer(forKey: Self.chipCountKey)
        if savedChipCount > 0 {
            initialChipCount = savedChipCount
        }

        delegate?.didLoadSavedSettings()
        validateForm()
    }

    private func saveSettings() {
        userDefaults.set(playerNames, forKey: Self.playerNamesKey)
        userDefaults.set(initialChipCount, forKey: Self.chipCountKey)
    }

    func resetToDefaults() {
        userDefaults.removeObject(forKey: Self.playerNamesKey)
        userDefaults.removeObject(forKey: Self.chipCountKey)

        playerNames = ["", ""]
        initialChipCount = 100
        playerCount = 2

        delegate?.didLoadSavedSettings()
        validateForm()
    }

    func incrementPlayerCount() {
        guard playerCount < maxPlayerCount else { return }
        playerNames.append("")
        playerCount += 1
    }

    func decrementPlayerCount() {
        guard playerCount > minPlayerCount else { return }
        playerNames.removeLast()
        playerCount -= 1
    }

    func updatePlayerName(at index: Int, name: String) {
        guard index < playerNames.count else { return }
        playerNames[index] = name
        validateForm()
    }

    func updateInitialChipCount(_ count: Int) {
        initialChipCount = max(1, count)
        validateForm()
    }

    func createGameSession() -> GameSession {
        saveSettings()
        let session = createGameSessionUseCase.execute(
            playerNames: playerNames,
            initialChipCount: initialChipCount
        )
        gameRepository.saveSession(session)
        return session
    }

    private func validateForm() {
        let allNamesValid = playerNames.allSatisfy { !$0.trimmingCharacters(in: .whitespaces).isEmpty }
        let chipCountValid = initialChipCount > 0
        delegate?.didValidateForm(allNamesValid && chipCountValid)
    }
}

File: ChipTallyAOS/app/src/main/java/com/chiptally/presentation/setup/SetupViewModel.kt
package com.chiptally.presentation.setup

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.chiptally.data.repository.GameRepositoryImpl
import com.chiptally.domain.model.GameSession
import com.chiptally.domain.usecase.CreateGameSessionUseCase

class SetupViewModel(
    private val createGameSessionUseCase: CreateGameSessionUseCase = CreateGameSessionUseCase(),
    private var gameRepository: GameRepositoryImpl? = null
) : ViewModel() {

    private val _playerCount = MutableLiveData(2)
    val playerCount: LiveData<Int> = _playerCount

    private val _playerNames = MutableLiveData(mutableListOf("", ""))
    val playerNames: LiveData<MutableList<String>> = _playerNames

    private val _initialChipCount = MutableLiveData(100)
    val initialChipCount: LiveData<Int> = _initialChipCount

    private val _isFormValid = MutableLiveData(false)
    val isFormValid: LiveData<Boolean> = _isFormValid

    private val _gameSession = MutableLiveData<GameSession?>()
    val gameSession: LiveData<GameSession?> = _gameSession

    val minPlayerCount = 2
    val maxPlayerCount = 10

    fun setRepository(repository: GameRepositoryImpl) {
        gameRepository = repository
    }

    fun loadSavedSettings() {
        gameRepository?.let { repo ->
            val savedNames = repo.loadPlayerNames()
            if (!savedNames.isNullOrEmpty()) {
                _playerNames.value = savedNames.toMutableList()
                _playerCount.value = savedNames.size
            }

            val savedChipCount = repo.loadChipCount()
            if (savedChipCount > 0) {
                _initialChipCount.value = savedChipCount
            }
        }
        validateForm()
    }

    fun incrementPlayerCount() {
        val current = _playerCount.value ?: minPlayerCount
        if (current < maxPlayerCount) {
            _playerNames.value?.add("")
            _playerCount.value = current + 1
            validateForm()
        }
    }

    fun decrementPlayerCount() {
        val current = _playerCount.value ?: minPlayerCount
        if (current > minPlayerCount) {
            _playerNames.value?.removeAt(_playerNames.value!!.lastIndex)
            _playerCount.value = current - 1
            validateForm()
        }
    }

    fun updatePlayerName(index: Int, name: String) {
        _playerNames.value?.let { names ->
            if (index < names.size) {
                names[index] = name
                _playerNames.value = names
                validateForm()
            }
        }
    }

    fun updateInitialChipCount(count: Int) {
        _initialChipCount.value = maxOf(1, count)
        validateForm()
    }

    fun createGameSession() {
        val names = _playerNames.value ?: return
        val chipCount = _initialChipCount.value ?: 100

        gameRepository?.savePlayerNames(names)
        gameRepository?.saveChipCount(chipCount)

        val session = createGameSessionUseCase.execute(names, chipCount)
        gameRepository?.saveSession(session)
        _gameSession.value = session
    }

    fun resetToDefaults() {
        gameRepository?.clearSettings()
        _playerNames.value = mutableListOf("", "")
        _initialChipCount.value = 100
        _playerCount.value = 2
        validateForm()
    }

    private fun validateForm() {
        val names = _playerNames.value ?: return
        val chipCount = _initialChipCount.value ?: 0

        val allNamesValid = names.all { it.trim().isNotEmpty() }
        val chipCountValid = chipCount > 0

        _isFormValid.value = allNamesValid && chipCountValid
    }
}
